File: src/main/java/io/github/freya022/botcommands/api/parameters/ParameterResolverFactory.java
package io.github.freya022.botcommands.api.parameters;

import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService;
import io.github.freya022.botcommands.api.core.utils.ReflectionKt;
import io.github.freya022.botcommands.api.core.utils.ReflectionUtils;
import kotlin.NotImplementedError;
import kotlin.reflect.KClass;
import org.jetbrains.annotations.NotNull;

/**
 * Factory for {@link ParameterResolver parameter resolvers}.
 *
 * @see ParameterResolver
 */
@InterfacedService(acceptMultiple = true)
public abstract class ParameterResolverFactory<T extends ParameterResolver<?, R>, R> {
	private final KClass<T> resolverType;
	private final KClass<R> jvmErasure;

	/**
	 * Constructs a new parameter resolver factory
	 *
	 * @param resolverType Class of the resolver
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolverFactory(Class<T> resolverType, @NotNull Class<R> clazz) {
		this.resolverType = ReflectionUtils.toKotlin(resolverType);
		this.jvmErasure = ReflectionUtils.toKotlin(clazz);
	}

	/**
	 * Constructs a new parameter resolver factory
	 *
	 * @param resolverType Class of the resolver
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolverFactory(KClass<T> resolverType, @NotNull KClass<R> clazz) {
		this.resolverType = resolverType;
		this.jvmErasure = clazz;
	}

	public KClass<T> getResolverType() {
		return resolverType;
	}

	public KClass<R> getJvmErasure() {
		return jvmErasure;
	}

	@NotNull
	public abstract T get(@NotNull ParameterWrapper parameter);

	@Override
	public String toString() {
		return "ParameterResolverFactory(resolverType=" + ReflectionKt.getShortQualifiedName(resolverType) + ", jvmErasure=" + ReflectionKt.getSimpleNestedName(jvmErasure) + ")";
	}

	@SuppressWarnings({"unchecked", "rawtypes"})
	public static <T extends ParameterResolver<?, R>, R> ParameterResolverFactory<T, R> singleton(T resolver) {
		throw new NotImplementedError();
//		return new ParameterResolverFactory<>(resolver.getClass(), ReflectionUtils.toJava(resolver.getJvmErasure())) {
//			@NotNull
//			@Override
//			public T get(@NotNull ParameterWrapper parameter) {
//				return resolver;
//			}
//
//			@Override
//			public String toString() {
//				return "SingletonParameterResolverFactory(resolver=" + resolver + ")";
//			}
//		};
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/parameters/ParameterResolverFactory.kt
package io.github.freya022.botcommands.api.parameters

import io.github.freya022.botcommands.api.core.reflect.ParameterWrapper
import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService
import io.github.freya022.botcommands.api.core.service.annotations.ResolverFactory
import io.github.freya022.botcommands.api.core.utils.shortQualifiedName
import io.github.freya022.botcommands.api.parameters.resolvers.IParameterResolver
import kotlin.reflect.KClass
import kotlin.reflect.KType

/**
 * Factory for [parameter resolvers][ParameterResolver].
 *
 * Your implementation needs to be annotated with [@ResolverFactory][ResolverFactory].
 *
 * ### How it works
 *
 * A factory determines if a given parameter is supported, if so, a parameter resolver will be created.
 *
 * **Note:** If multiple factories return `true` in [isResolvable] for a given type, an exception is thrown.
 *
 * ### Use cases
 *
 * This is particularly useful if your parameter's type has generics, which you may read in [ParameterWrapper.type],
 * or if you need to read an annotation on the [parameter][ParameterWrapper.parameter]
 * to adjust the resolver's behavior, for example.
 *
 * In case you want to read generics, you can read them off the [KType] in [ParameterWrapper],
 * but you can only do that with Kotlin.
 *
 * In case you want to read the annotations, you can use the methods supplied by [ParameterWrapper].
 *
 * @param T Type of the returned parameter resolver
 *
 * @see TypedParameterResolverFactory
 * @see ParameterResolver
 */
@InterfacedService(acceptMultiple = true)
abstract class ParameterResolverFactory<T : IParameterResolver<T>>(val resolverType: KClass<out T>) {
    constructor(resolverType: Class<out T>) : this(resolverType.kotlin)

    /**
     * List of types as strings that are supported by this resolver factory.
     *
     * This should be the types of what the returned parameter resolvers are capable of returning.
     *
     * This is only used for logging purposes.
     */
    abstract val supportedTypesStr: List<String>

    /**
     * Determines if a given parameter is supported, only one factory must return `true`.
     *
     * This only gets called if the requested resolver is compatible with [resolverType].
     */
    abstract fun isResolvable(request: ResolverRequest): Boolean

    /**
     * Returns a [ParameterResolver] for the given parameter.
     *
     * This is only called if [isResolvable] returned `true`.
     */
    abstract fun get(request: ResolverRequest): T

    override fun toString(): String {
        return "ParameterResolverFactory(resolverType=${resolverType.shortQualifiedName})"
    }
}